File: LeetCode/Python/leetcode_2106_Maximum_Fruits_Harvested.py
# LeetCode #2106 - Maximum Fruits Harvested After at Most K Steps
# Date: 2025-08-03

from bisect import bisect_left, bisect_right
from typing import List

class Solution:
    def maxTotalFruits(self, fruits: List[List[int]], startPos: int, k: int) -> int:
        n = len(fruits)

        indices = [0] * n
        preSum = [0] * n

        for i in range(n):
            indices[i] = fruits[i][0]
            preSum[i] = fruits[i][1] + (preSum[i - 1] if i > 0 else 0)

        maxfruits = 0

        for d in range((k // 2) + 1):
            i = startPos - d
            remain = k - 2 * d
            j = startPos + remain

            left = bisect_left(indices, i)
            right = bisect_right(indices, j) - 1

            if left <= right:
                total = preSum[right] - (preSum[left - 1] if left > 0 else 0)
                maxfruits = max(total, maxfruits)

            j = startPos + d
            i = startPos - remain

            left = bisect_left(indices, i)
            right = bisect_right(indices, j) - 1

            if left <= right:
                total = preSum[right] - (preSum[left - 1] if left > 0 else 0)
                maxfruits = max(total, maxfruits)

        return maxfruits

File: LeetCode/C++/leetcode_2106_Maximum_Fruits_Harvested.cpp
// LeetCode #2106 - Maximum Fruits Harvested After at Most K Steps
// Date: 2025-08-03

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxTotalFruits(vector<vector<int>>& fruits, int startPos, int k) {
        int n = fruits.size();
        vector<int> indices(n);
        vector<int> preSum(n);

        for (int i = 0; i < n; i++) {
            indices[i] = fruits[i][0];
            preSum[i] = fruits[i][1] + (i > 0 ? preSum[i - 1] : 0);
        }

        int maxfruits = 0;

        for (int d = 0; d <= k / 2; d++) {
            int i = startPos - d;
            int remain = k - 2 * d;
            int j = startPos + remain;

            int left = lower_bound(indices.begin(), indices.end(), i) - indices.begin();
            int right = upper_bound(indices.begin(), indices.end(), j) - indices.begin() - 1;

            if (left <= right) {
                int total = preSum[right] - (left > 0 ? preSum[left - 1] : 0);
                maxfruits = max(maxfruits, total);
            }

            j = startPos + d;
            i = startPos - (k - 2 * d);

            left = lower_bound(indices.begin(), indices.end(), i) - indices.begin();
            right = upper_bound(indices.begin(), indices.end(), j) - indices.begin() - 1;

            if (left <= right) {
                int total = preSum[right] - (left > 0 ? preSum[left - 1] : 0);
                maxfruits = max(maxfruits, total);
            }
        }

        return maxfruits;
    }
};
